File: java/853.car-fleet.java
/*
 * @lc app=leetcode id=853 lang=java
 *
 * [853] Car Fleet
 */

// @lc code=start
class Solution {
    public int carFleet(int target, int[] position, int[] speed) {
        Integer[] idx = new Integer[position.length];

        for(int i = 0; i < idx.length; i++) {
            idx[i] = i;
        }

        Arrays.sort(idx, Comparator.comparingInt(i -> position[i]));

        Deque<Double> stk = new ArrayDeque();

        for(int i : idx) {
            double arrival = (double) (target - position[i]) / speed[i];

            while(!stk.isEmpty() && arrival >= stk.peek()) {
                stk.pop();
            }

            stk.push(arrival);
        }

        return stk.size();
    }
}
// @lc code=end

File: cpp/853.car-fleet.cpp
/*
 * @lc app=leetcode id=853 lang=cpp
 *
 * [853] Car Fleet
 */

// @lc code=start
class Solution {
public:
    int carFleet(int target, vector<int>& position, vector<int>& speed) {
        int n = position.size();
        vector<int> idx(n);

        for(int i = 0; i < n; i++) {
            idx[i] = i;
        }

        std::sort(idx.begin(), idx.end(), [&](int a, int b) -> bool {
            return position[a] < position[b];
        });

        std::stack<double> stk;

        for(int i = 0; i < n; i++) {
            double arrival = static_cast<double>(target - position[idx[i]]) / speed[idx[i]];
            while(!stk.empty() && arrival >= stk.top()) {
                stk.pop();
            }
            stk.push(arrival);
        }

        return stk.size();
    }
};
// @lc code=end

